### PR TITLE
ApplyCalibration should handle bar scans

### LINKED ISSUE #27876: ApplyCalibration should handle bar scans
Algorithm ApplyCalibration should be able to handle variable number of columns to account for different types of calibration
Common column:
- pixel ID

Current calibration:
- pixel V3D position

Barscan calibration:
- pixel Y position
- pixel height

Apparent tube calibration:
- pixel width

### PR DESCRIPTION
**Pull request by jmborr**
**Description of work.**
- [x] Algorithm accepts variable number of columns.
- [x] Algorithm accepts columns for pixel Y-coordinate, height, and width.
- [x] Unit test involving pixel Y-coordinate, height, and width.
- [x] Update release notes.
- [x] Update algorithm's documentation, including the doctest

**To test:**
Run the following script which will modify the detector's heights of GPSANS. The script will also generate a workspace named `heights` which contains these modified heights as intentisity values, so that one can visualize them in the instrument view.
The script uses the deprecated property `PositionTable` to verify that a deprecation notice is sent to the Results Log widget.

```
import time
# Resize the detector heights of GPSANS
# We double the height of detectors with even detector ID, and we half the height for the rest
ws = LoadEmptyInstrument(InstrumentName='GPSANS', OutputWorkspace='ws')
compInfo = ws.componentInfo()
detInfo = ws.detectorInfo()

# Create and populate the calibration table
calibTable = CreateEmptyTableWorkspace(OutputWorkspace='CalibTable')
calibTable.addColumn(type='int',name='Detector ID')
calibTable.addColumn(type='double',name='Detector Height')
for i in range(ws.getNumberHistograms()):
    det = ws.getDetector(i)
    detector_id = det.getID()
    compInfo_index = detInfo.indexOf(detector_id)
    if detInfo.isMonitor(compInfo_index) is True:
        continue
    factor = 2 if detector_id % 2 == 0 else 0.5
    height = factor * compInfo.shape(compInfo_index).getBoundingBox().width().Y()
    nextRow = {'Detector ID': detector_id,  'Detector Height': height}
    calibTable.addRow(nextRow)

# Apply the calibration and find out how long it takes
start_time = time.time()
ApplyCalibration(Workspace='ws', PositionTable='CalibTable')
elapsed_time = time.time() - start_time
print('ApplyCalibration took ' + str(int(1000 * elapsed_time)) + ' mili-seconds')

# Visualize the calibration by using the detector heights as the intensities. First we need to collect the new detector heights
ws = mtd['ws']
compInfo = ws.componentInfo()
detInfo = ws.detectorInfo()
heights = list()
for i in range(ws.getNumberHistograms()):
    det = ws.getDetector(i)
    compInfo_index = detInfo.indexOf(det.getID())
    if detInfo.isMonitor(compInfo_index) is True:
        continue
    scaling_y = compInfo.scaleFactor(compInfo_index).Y()
    heights.append(scaling_y * compInfo.shape(compInfo_index).getBoundingBox().width().Y())

# Create a workspace using the heights as intensities
CreateWorkspace(DataX=[0, 1.], DataY=heights, NSpec=len(heights), OutputWorkspace='heights')
LoadInstrument(Workspace='heights', RewriteSpectraMap=True, InstrumentName='GPSANS')
```
Typical output of the script to `STDOUT`:

> Wed Feb 12 14:16:28 2020: Script execution started.
> ApplyCalibration took 82 mili-seconds
> Wed Feb 12 14:16:45 2020: Script execution finished.

And a snapshot of the instrument view of workspace `heights`

Fixes #27876

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: docs/source/algorithms/LoadRaw-v3.rst [hunk 1]
```diff
@@ -97,8 +97,8 @@ Output:
 
   Workspace has 2890 spectra
   Is detector 1001 a monitor? True
-  Workspace contains 35 logs
-  Workspace has property INT1: True
+  Workspace contains 33 logs
```
Comment (AndreiSavici) [inline docs/source/algorithms/LoadRaw-v3.rst]: This does not make any sense. Why did the numbers of logs change?

Comment (peterfpeterson) [reply docs/source/algorithms/LoadRaw-v3.rst]: What makes even less sense is that switching back to `master` gets 33 again.

File: Testing/Data/UnitTest/HRP39180.RAW.md5 [hunk 2]
```diff
@@ -0,0 +1 @@
+7d19937fbfb68e962c190454a128fde6
```
Comment (martyngigg) [inline Testing/Data/UnitTest/HRP39180.RAW.md5]: Can we avoid adding a new copy of this file hash? We'll now have the same file in `UnitTest`, `DocTest` and `SystemTest` if we add it here and if it ever needs to be updated we have to remember it is in 3 places.

Unit tests/ system tests have access to DocTest so can we just leave the copy that is in there and remove this and the one in `Data/SystemTest?`

I think this has also caused the seemingly strange failure of the doctests for LoadRaw3. The `UnitTest` directory is first in the search path so this file would now be picked up. The RAW file format does not have all log data contained within the file but relies on separate `*.txt` files alongside the `.raw` to load additional log data. In `Data/DocTest` there are 3 files related to HRP39180:
```
ExternalData/Testing/Data/DocTest/HRP39180_INT1.txt
ExternalData/Testing/Data/DocTest/HRP39180_INT2.txt
ExternalData/Testing/Data/DocTest/HRP39180.RAW
```
The .raw file gives 33 logs and then the 2 extra .txt files give bring the total to 35. By copying this file to UnitTest but without the logs it now doesn't see the extra `.txt` files so gives you back only 33. When you switch back to master you will still have the file in the downloaded 'ExternalData/Testing/Data/UnitTest' directory so master will seem like it fails too. For this reason we always remove the links to the downloaded ExternalData files on the [buildservers](https://github.com/mantidproject/mantid/blob/master/buildconfig/Jenkins/buildscript#L200). If you do the same then master should start working again.

The upshot here is I think if we remove this file from `UnitTest` and ``SystemTest` then I think that will be fine.

**Review by martyngigg — DISMISSED** [review]

File: Testing/Data/UnitTest/HRP39180.RAW.md5 (hunk 2 above)
Comment (AndreiSavici) [reply Testing/Data/UnitTest/HRP39180.RAW.md5]: :+1:

**Review by AndreiSavici — APPROVED** [review]

**Comment by AndreiSavici** [comment]
Note that the detector size in the instrument view does not change

File: Testing/Data/UnitTest/HRP39180.RAW.md5 (hunk 2 above)
Comment (jmborr) [reply Testing/Data/UnitTest/HRP39180.RAW.md5]: @martyngigg it worked. Thanks a million

**Comment by martyngigg** [comment]
@AndreiSavici I think this is a known "feature" of the instrument view. It optimizes during tubes as a single cylinder with a texture map for the coloring but the texture map is not able to show the different pixel sizes. They are nevertheless correct in the underlying instrument.

**Pull request merged**